File: backend/app/core/api/reset.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from app.api.dependencies import get_current_user, get_data_reset_service
from app.core.entities.system import User
from app.core.exceptions import ValidationError
from app.core.redis import consume_backup_receipt
from app.core.services.data_reset import SCOPES, DataResetService

router = APIRouter()


class ResetRequest(BaseModel):
    scopes: list[str] = Field(
        ...,
        description=f"One or more of: {list(SCOPES)}. Pass all five for a full wipe.",
        min_length=1,
    )
    backup_receipt: str = Field(
        ...,
        description=(
            "The receipt returned by GET /portfolio/backup (X-Backup-Receipt response "
            "header). Single-use and expires 10 minutes after export — re-export if it's "
            "been longer, or if data has changed since."
        ),
    )


@router.get("/reset/scopes")
def list_reset_scopes():
    return {"scopes": list(SCOPES)}


@router.get("/reset/preview")
def preview_data_reset(
    scopes: str = Query(
        ...,
        description=f"Comma-separated list of scopes to preview. One or more of: {list(SCOPES)}.",
    ),
    user: User = Depends(get_current_user),
    service: DataResetService = Depends(get_data_reset_service),
):
    """Read-only row counts per scope for a confirmation screen — no deletion,
    no backup-receipt requirement (nothing destructive happens here). Rejects
    while a reset is actually running, since counts would be mid-flux."""
    scope_list = [s.strip() for s in scopes.split(",") if s.strip()]
    try:
        counts = service.preview(scope_list, owner_id=user.id)
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return {"counts": counts}


@router.post("/reset")
def run_data_reset(
    body: ResetRequest,
    user: User = Depends(get_current_user),
    service: DataResetService = Depends(get_data_reset_service),
):
    # Validate scopes before consuming the (single-use) receipt — a typo in the
    # request body shouldn't burn a valid backup and force a re-export.
    unknown = set(body.scopes) - set(SCOPES)
    if unknown:
        raise HTTPException(status_code=400, detail=f"Unknown reset scope(s): {sorted(unknown)}")

    if not consume_backup_receipt(body.backup_receipt):
        raise HTTPException(
            status_code=409,
            detail=(
                "No valid, unexpired backup found for this receipt. Export a fresh backup "
                "via GET /portfolio/backup and use its X-Backup-Receipt header before "
                "resetting."
            ),
        )

    try:
        results = service.reset(body.scopes, owner_id=user.id, actor_id=user.id)
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))

    return {"status": "success", "cleared": results}

File: backend/src/routes/settings/reset.ts
import { Router } from "express";
import { SCOPES, previewReset, runReset } from "../../lib/settings/dataReset";
import { consumeBackupReceipt } from "../../lib/settings/resetRedis";
import { getCurrentUser } from "../../lib/users";
import { ConflictError, ValidationError } from "../../lib/errors";

export const resetRouter = Router();

resetRouter.get("/reset/scopes", (_req, res) => {
  res.json({ scopes: SCOPES });
});

resetRouter.get("/reset/preview", async (req, res) => {
  const user = await getCurrentUser();
  const scopesParam = String(req.query.scopes ?? "");
  const scopeList = scopesParam.split(",").map((s) => s.trim()).filter(Boolean);
  const counts = await previewReset(scopeList, user.id);
  res.json({ counts });
});

resetRouter.post("/reset", async (req, res) => {
  const user = await getCurrentUser();
  const { scopes, backup_receipt: backupReceipt } = req.body ?? {};

  // Validate scopes BEFORE consuming the single-use receipt — a typo in the
  // request body shouldn't burn a valid backup and force a re-export. This
  // also rejects a missing/empty/non-array `scopes` here rather than letting
  // it reach consumeBackupReceipt (which would burn the receipt) and then
  // runReset's internal validateScopes (400 for empty, uncaught 500 for
  // non-array since it calls .filter() directly).
  if (!Array.isArray(scopes) || scopes.length === 0) {
    throw new ValidationError("scopes must be a non-empty array");
  }
  const unknown = scopes.filter((s: string) => !SCOPES.includes(s as never));
  if (unknown.length > 0) {
    // Python's route re-raises this as HTTPException(400), matching
    // errorHandler.ts's ValidationError -> 400 mapping — not
    // RequestValidationError, which maps to 422.
    throw new ValidationError(`Unknown reset scope(s): ${JSON.stringify(unknown.sort())}`);
  }

  if (!(await consumeBackupReceipt(backupReceipt))) {
    throw new ConflictError(
      "No valid, unexpired backup found for this receipt. Export a fresh backup via GET /portfolio/backup and use its X-Backup-Receipt header before resetting.",
    );
  }

  const results = await runReset(scopes, user.id, user.id);
  res.json({ status: "success", cleared: results });
});
